### PR TITLE
Joealsubash/cppbutler

### LINKED ISSUE #23: migrate hfButler to use C++ classes
The command line tool for accessing TPG firmware, hfButler, is currently standalone Python.  This task is to migrate it to puse the C++, while maintaining the same user interface, so that both CLI and DAQ app use the same underlying C++ code.

### PR DESCRIPTION
**Pull request by joealsubash**
hfButler_cpp_clone.py will eventually become hfButler.py after it has been tested by users in-situ. The old butler still exists to allow users to revert back incase the cpp version crashes/behaves-unexpectedly.

### CODE REVIEW

File: scripts/hfButler_cpp_clone.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        if dpr_mux:
+            dprNode = ln.get_data_router_node().get_dpr_node()
+            print("Configuring DPR mux : ", dpr_mux)
+            dprNode.set_mux_in(dpr_mux_choices[dpr_mux][0], True)
+            dprNode.set_mux_out(dpr_mux_choices[dpr_mux][1], True)
+        
+        if dr_on is not None:
+            print("Configuring data-reception : ", str(dr_on))
+            drNode = ln.get_data_router_node().get_data_reception_node()
+            drNode.getNode("csr.ctrl.en").write(dr_on)
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py]: DataReceptionNode::enable() would be better

File: scripts/hfButler_cpp_clone.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+            drNode.getClient().dispatch()
+
+        if drop_empty is not None:
+            strmArrayNode = ln.get_stream_proc_array_node()
+            strmNode = strmArrayNode.get_stream_proc_node()
+
+            for i in range(obj.mConfigInfo['n_port']):
+                print("Setting drop empty to {} on stream {} ".format(drop_empty, i))
+                strmArrayNode.stream_select(i)
+                strmNode.getNode('csr.ctrl.drop_empty').write(drop_empty)
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py]: StreamProcessorNode::drop_empty() would be better

File: scripts/hfButler_cpp_clone.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        strmNode = strmArrayNode.get_stream_proc_node()
+
+        print('Configuring channel masks in pipelines ', ', '.join((str(p) for p in pipes)))
+        mask_w = 0x0
+        for b in chans:
+            mask_w |= (1 << b)
+
+        for p in pipes:
+            strmArrayNode.stream_select(p)
+            strmNode.set_channel_mask_all(mask_w)
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py:145]: this sets the same mask on all pipelines, right?  i'm not sure this is desirable.  we should check this replicates old hfButler behaviour, at least.

File: scripts/hfButler_cpp_clone.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+def cr_if(obj, cr_on, drop_empty):
+
+    crNode = obj.podNode.get_crif_node()
+    if cr_on is not None:
+        crNode.getNode('csr.ctrl.en').write(0x01)
+        crNode.getNode('csr.ctrl.drop_empty').write(0x01)
+        obj.mHW.dispatch()
+        
+    if drop_empty is not None:
+        crNode.getNode('csr.ctrl.drop_empty').write(0x01)
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py]: replace with CentralRouterInterfaceNode::set_drop_empty()

File: scripts/hfButler_cpp_clone.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+# -----------------------------------------------------------------------------
+@cli.command('cr-if')
+@click.option('--on/--off', 'cr_on', help='Enable central-router interface block', default=None)
+@click.option('--drop-empty/--keep-empty', 'drop_empty', help='Drop empty hit packets', default=None)
+@click.pass_obj
+def cr_if(obj, cr_on, drop_empty):
+
+    crNode = obj.podNode.get_crif_node()
+    if cr_on is not None:
+        crNode.getNode('csr.ctrl.en').write(0x01)
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py]: replace with CentralRouterInterfaceNode::enable()

File: scripts/hfButler_cpp_clone.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+@cli.command('capture-sink')
+@click.option('--path', type=click.Path(exists=False), help='Capture file path', default='b33_dump.txt')
+@click.option('--timeout', type=float, help='Capture timeout', default=10)
+@click.option('--drop-idles', is_flag=True, help='Drop idle words', default=None)
+@click.pass_obj
+def capture_sink(obj, path, timeout, drop_idles):
+
+    osNode = obj.podNode.get_output_sink_node()
+
+    print('Capturing axis32bsink data')
```
Comment (jimbrooke) [inline scripts/hfButler_cpp_clone.py:328]: not necessary immediately, but this should all be replaced with appropriate methods on OutputSinkNode

File: scripts/hfButler_cpp_clone.py (hunk 1 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py]: Change made.

File: scripts/hfButler_cpp_clone.py (hunk 2 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py]: Change made

File: scripts/hfButler_cpp_clone.py (hunk 3 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py:145]: Change made

File: scripts/hfButler_cpp_clone.py (hunk 4 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py]: Change made

File: scripts/hfButler_cpp_clone.py (hunk 5 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py]: Change made

File: scripts/hfButler_cpp_clone.py (hunk 6 above)
Comment (joealsubash) [reply scripts/hfButler_cpp_clone.py:328]: Done reasonably I think. Please review.

Comment (jimbrooke) [reply scripts/hfButler_cpp_clone.py:328]: LGTM

**Pull request merged**